Handle new media upload errors

daniloercoli opened the pull request:
Fixes #6222 by handling `CONNECTION_ERROR`, `EXCEEDS_FILESIZE_LIMIT` and `EXCEEDS_MEMORY_LIMIT` during media uploading. 
There is now a single routine that from a given `MediaError` returns the res ID of the error message. It's going to be used in the Editor, Media Library, and post upload service (mostly from legacy editor).

For reference/details about when these new errors are triggered see the original ticket.

cc @nbradbury for helping about wording these new error strings.

nbradbury commented:
**media_error_http_too_large_upload**
The string says to try enabling Optimize Images, but there's no logic to detect whether that's already enabled and we don't know if they were uploading an image. Also, rather than use "Site->Settings" I think it's better to say "in your site's settings" (the reason being that the user may not know that "Site->" refers to the first page in the masterbar, since the tabs are unlabeled).

**media_error_exceeds_php_filesize** and **media_error_exceeds_memory_limit**
Both of these are quite long for toasts, and they're very geeky. Is it reasonable to expect the user to know how to change php.ini? If we have no alternative but to mention php.ini, perhaps we should add something to the Helpshift FAQ about it.

**media_error_generic_connection_error**
We have a number of strings that basically say "connection error" or "no connection." It would be nice to consolidate these and use a single string (like `no_network_title` or `no_network_message`)

nbradbury commented:
This is beyond the scope of this PR, but it would be nice if we could *help* the user rather than simply display a toast. For example, if the user uploads an image that's too large, show a dialog explaining the problem and ask them if they'd like to try uploading a reduced (optimized) version of the image. At the bottom of the dialog there could be a checkbox saying something like, "Always optimize images."

daniloercoli commented:
> This is beyond the scope of this PR, but it would be nice if we could help the user rather than simply display a toast. For example, if the user uploads an image that's too large, show a dialog explaining the problem and ask them if they'd like to try uploading a reduced (optimized) version of the image. At the bottom of the dialog there could be a checkbox saying something like, "Always optimize images."

I completely agree with you on this. We should already have tickets filled in the Media project about it.

daniloercoli commented:
@nbradbury This is ready for another pass.

1. I've added the check to be sure `optimize image` is OFF before suggesting to enable it. Also changed the wording as you've suggested.
2. `media_error_exceeds_php_filesize` and `media_error_exceeds_memory_limit` are now less geeky. the first is the same error message WordPress displays on the web.
We should add Helpshift FAQs for both of them. cc @rachelmcr

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,51 @@ public void onCancel(DialogInterface dialog) {
+     * @param error The media error occurred
+     * @return String  The associated error message.
+     */
+    public static String getErrorMessage(final Context context, boolean suggestMediaOptimization, final MediaModel media, final MediaStore.MediaError error) {
```
The parameter `suggestMediaOptimization` seems like a misnomer (and I not sure the logic below is correct) since what's being passed is whether image optimization is enabled for the site. Can we rename this `isImageOptimizationEnabled`?

daniloercoli replied:
My first attempt was with `isImageOptimizationEnabled`, but it will be soon used for videos too. So i switched to `suggestMediaOptimization`. That means in case of error to suggest to the user to enable video or image optimization.

nbradbury replied:
That logic looks incorrect, though. For example, [here](https://github.com/wordpress-mobile/WordPress-Android/pull/6285/commits/3316c171fcb2b63ee934050ebf06eed0dbc278eb#diff-fdc9b5c706ac889526b8df93c55351ceR635) we're passing whether image optimization is enabled, so we're only going to suggest optimization if it's already enabled. Or am I reading that wrong?

daniloercoli replied:
You're right :)

daniloercoli replied:
It should be fixed now. :crossed_fingers:

nbradbury commented:
Hmmm, I'm wondering now if we're over-complicating things. What do you think about dropping the optimization parameters altogether, then changing the message to state the user might want to adjust their optimization settings. 

That way we don't have to pass anything about optimization and there aren't two different messages depending on whether optimization is enabled. And it makes more sense anyway since it's possible for the user to have optimization enabled but the chosen size is still too large.

daniloercoli commented:
Simplicity wins! 👍 

Let me know if the new message is OK, or we need to change it.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/posts/services/PostUploadService.java:
```diff
@@ -580,7 +580,7 @@ private String uploadImageFile(MediaFile mediaFile, SiteModel site) {
 
     private @NonNull String getErrorMessageFromMediaError(OnMediaUploaded event) {
         // FIXME : We set isImageOptimizationEnabled and isVideoOptimizationEnabled to true because we don't want display the hint about them.
```
This comment can be dropped.

nbradbury commented on WordPress/src/main/res/values-en-rAU/strings.xml:
```diff
@@ -66,7 +66,7 @@ Language: en_AU
 	<string name="media_error_timeout">Server response timed out, you may have a poor connection or server error</string>
 	<string name="media_error_internal_server_error">Upload error. Try enabling Optimise Images in Site->Settings</string>
-	<string name="media_error_too_large_upload">File too large to upload. Try enabling Optimise Images in Site->Settings</string>
+	<string name="media_error_http_too_large_upload">File too large to upload. Try enabling Optimise Images in Site->Settings</string>
```
Looks like these non-English strings need to be updated to match the English strings from the last commit.

daniloercoli replied:
I think we can skip this, since the non-English strings will be updated before the release by the automated scripts.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,47 @@ public void onCancel(DialogInterface dialog) {
+                if (media.isVideo()) {
+                    return context.getString(R.string.media_error_http_too_large_video_upload);
+                } else {
+                    return context.getString(R.string.media_error_http_too_large_photo_upload) + ". " +
```
Why are we concatenating two strings here when a single string resource would work?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,47 @@ public void onCancel(DialogInterface dialog) {
+            case TIMEOUT:
+                return context.getString(R.string.media_error_timeout);
+            case CONNECTION_ERROR:
+                return context.getString(R.string.connection_error) + ". "
```
I think we can drop `media_error_generic_connection_error` and just use `connection_error` here.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,47 @@ public void onCancel(DialogInterface dialog) {
+                            context.getString(R.string.media_error_suggest_optimize_image);
+                }
+            case SERVER_ERROR:
+                return context.getString(R.string.media_error_internal_server_error);
```
This message says to try enabling optimize images, but again it may already be enabled. Also, how likely is it that image optimization would fix the problem? I don't want to send users on a wild goose chase.

daniloercoli replied:
We've seen that most of the times internal server errors are due to memory problem uploading pictures. The server may have other problems too. Should we remove the optimize image section of it?

nbradbury replied:
If there's a decent chance that the error is due to upload size, then it makes sense to keep the optimize images part of the message.

daniloercoli replied:
Yes, I was able to reproduce it by sending larger images to self hosted testing sites of mine, and got back the internal server error - This is due to the PHP memory exhausted underlaying issue. I bet all depends on the conf on the sever. Better to keep it for now.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,47 @@ public void onCancel(DialogInterface dialog) {
+            case SERVER_ERROR:
+                return context.getString(R.string.media_error_internal_server_error);
+            case TIMEOUT:
+                return context.getString(R.string.media_error_timeout);
```
I think we can change this to "Server took too long to respond."

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,47 @@ public void onCancel(DialogInterface dialog) {
+
+        switch (error.type) {
+            case FS_READ_PERMISSION_DENIED:
+                return context.getString(R.string.error_media_insufficient_fs_permissions);
```
This error message - "Read permission required on media file" - seems cryptic, but I'm not sure how to reword it. What would cause this to happen?

daniloercoli replied:
Not quite sure when this could happen. maybe if the user removed the read permission on storage after the first attempt to upload the media file failed.

nbradbury replied:
Mario confirmed that this refers to read permission on the device, so let's update the message to clarify that. How about "Read permission denied on device media" ?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java:
```diff
@@ -151,4 +153,47 @@ public void onCancel(DialogInterface dialog) {
+            case EXCEEDS_MEMORY_LIMIT:
+                return context.getString(R.string.media_error_exceeds_memory_limit);
+            case PARSE_ERROR:
+                return context.getString(R.string.error_media_parse_error);
```
"Could not parse response after media upload" is also quite cryptic. I'm assuming this means the server's response wasn't what we expected, and at this point we don't know whether the upload was successful? Or does it always mean the upload failed?

daniloercoli replied:
Unfortunately, we don't know whether the upload was successful or not in this case. In both cases although we can't add the media to the local app database, since we miss all of the metadata info about it. (ID, title, etc).

nbradbury replied:
In that case maybe a generic "Unexpected response from server" would work.

nbradbury commented:
:shipit:
